core/src/main/java/org/opensearch/sql/datasource/DataSourceServiceImpl.java

```diff
@@ -59,21 +61,36 @@ public DataSourceServiceImpl(Set<DataSourceFactory> dataSourceFactories,
   }
 
   @Override
-  public Set<DataSourceMetadata> getDataSourceMetadataSet() {
+  public Set<DataSourceMetadata> getDataSourceMetadataSet(boolean isDefaultDataSourceRequired) {
     List<DataSourceMetadata> dataSourceMetadataList
         = this.dataSourceMetadataStorage.getDataSourceMetadata();
     Set<DataSourceMetadata> dataSourceMetadataSet = new HashSet<>(dataSourceMetadataList);
-    dataSourceMetadataSet.add(DataSourceMetadata.defaultOpenSearchDataSourceMetadata());
+    if (isDefaultDataSourceRequired) {
+      dataSourceMetadataSet.add(DataSourceMetadata.defaultOpenSearchDataSourceMetadata());
+    }
+    removeAuthInfo(dataSourceMetadataSet);
     return dataSourceMetadataSet;
   }
 
+  @Override
+  public DataSourceMetadata getDataSourceMetadataSet(String datasourceName) {
+    Optional<DataSourceMetadata> dataSourceMetadataOptional
+        = this.dataSourceMetadataStorage.getDataSourceMetadata(datasourceName);
+    if (dataSourceMetadataOptional.isEmpty()) {
+      throw new IllegalArgumentException("DataSource with name: " + datasourceName
+          + " doesn't exist.");
+    }
+    removeAuthInfo(dataSourceMetadataOptional.get());
+    return dataSourceMetadataOptional.get();
+  }
```

Comment:
reuse getDataSourceMetadataSet()?

Reply:
I am querying the underlying index for datasource with particular name instead of fetching all and applying filter in the code.